File: Algorithms/144.BinaryTreePreorderTraversal.cpp
class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        vector<int> v;
        stack<TreeNode *> stack;
        TreeNode *p = root;
        while (!stack.empty() || p != NULL) {
            if (p != NULL) {
                v.push_back(p->val);
                if (p->right != NULL)
                    stack.push(p->right);
                p = p->left;
            }
            else {
                p = stack.top();
                stack.pop();
            }
        }
        return v;
    }

    /*
    // recursive version
    vector<int> preorderTraversal(TreeNode *root) {
        vector<int> v;
        if (root == NULL) {
            return v;
        }
        v.push_back(root->val);
        vector<int> left = preorderTraversal(root->left);
        for (int i=0; i < left.size(); i++) {
            v.push_back(left[i]);
        }
        vector<int> right = preorderTraversal(root->right);
        for (int i=0; i < right.size(); i++) {
            v.push_back(right[i]);
        }
        return v;
    }
    */
};

File: Algorithms/java/144.BinaryTreePreorderTraversal.java
public class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> v = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        TreeNode p = root;
        while (!stack.empty() || p != null) {
            if (p != null) {
                v.add(p.val);
                if (p.right != null)
                    stack.push(p.right);
                p = p.left;
            }
            else {
                p = stack.peek();
                stack.pop();
            }
        }
        return v;
    }
    /*
    // recursive version
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> v = new ArrayList<>();
        if (root == null) {
            return v;
        }
        v.add(root.val);
        List<Integer> left = preorderTraversal(root.left);
        for (int i=0; i < left.size(); i++) {
            v.add(left.get(i));
        }
        List<Integer> right = preorderTraversal(root.right);
        for (int i=0; i < right.size(); i++) {
            v.add(right.get(i));
        }
        return v;
    }
    */
}
